catroid/src/main/java/org/catrobat/catroid/io/DeviceUserDataAccessor.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.io;
+
+import android.support.annotation.VisibleForTesting;
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+import org.catrobat.catroid.content.Project;
+import org.catrobat.catroid.content.Scene;
+import org.catrobat.catroid.content.Sprite;
+import org.catrobat.catroid.formulaeditor.UserData;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+public abstract class DeviceUserDataAccessor {
+
+	public static final String TAG = DeviceUserDataAccessor.class.getSimpleName();
+	private File deviceFile;
+
+	public DeviceUserDataAccessor(File projectDirectory) {
+		deviceFile = new File(projectDirectory, getDeviceFileName());
+	}
+
+	protected abstract Object getLock();
+
+	@VisibleForTesting
+	public void setDeviceFile(File deviceFile) {
+		this.deviceFile = deviceFile;
+	}
+
+	public boolean readUserData(UserData userData) {
+
+		HashMap deviceMap = readMapFromJson();
+
+		if (deviceMap == null) {
+			userData.reset();
+			return false;
+		}
+
+		if (!deviceMap.containsKey(userData.getDeviceKey())) {
+			userData.reset();
+			return false;
+		}
+		userData.setValue(deviceMap.get(userData.getDeviceKey()));
+		return true;
+	}
+
+	public void writeUserData(UserData userData) throws IOException {
+		synchronized (getLock()) {
+			if (!deviceFile.exists()) {
```

Comment:
you disregard the return value of `createNewFile()`, here. so if it fails it can be very difficult to find out why it did not work because it then crashes somewhere during the json serialization. so I suggest you do something like this:
```
if (!deviceFile.exists() && !deviceFile.createNewFile()) {
	throw new IOException("Cannot create " + deviceFile.getAbsolutePath());
}
```